server/src/main/java/org/opensearch/search/aggregations/bucket/FastFilterRewriteHelper.java

```diff
@@ -439,61 +402,297 @@ public static boolean tryFastFilterAggregation(
+        throws IOException {
+        PointValues.Relation r = visitor.compare(pointTree.getMinPackedValue(), pointTree.getMaxPackedValue());
+
+        try {
+            switch (r) {
+                case CELL_INSIDE_QUERY:
+                    pointTree.visitDocIDs(visitor);
+                    debug.visitInner();
                     break;
+                case CELL_CROSSES_QUERY:
+                    if (pointTree.moveToChild()) {
+                        intersectWithRanges(visitor, pointTree, debug);
+                        pointTree.moveToSibling();
+                        intersectWithRanges(visitor, pointTree, debug);
+                        pointTree.moveToParent();
+                    } else {
+                        pointTree.visitDocValues(visitor);
+                        debug.visitLeaf();
+                    }
+                    break;
+                case CELL_OUTSIDE_QUERY:
+            }
+        } catch (CollectionTerminatedException e) {
+            logger.debug("Early terminate since no more range to collect");
+        }
+    }
+
+    /**
+     *
+     */
+    private static PointValues.IntersectVisitor getIntersectVisitor(RangeCollectorForPointTree collector) {
+        return new PointValues.IntersectVisitor() {
+            @Override
+            public void visit(int docID) throws IOException {
+                collector.count();
+            }
+
+            @Override
+            public void visit(int docID, byte[] packedValue) throws IOException {
+                long value = NumericUtils.sortableBytesToLong(packedValue, 0);
+                if (value > collector.activeRange[1]) {
+                    // need to move to next range
+                    collector.finalizePreviousRange();
+
+                    if (collector.iterateRangeEnd(value)) {
+                        throw new CollectionTerminatedException();
+                    }
+                }
+                if (collector.activeRange[0] <= value && value <= collector.activeRange[1]) {
+                    collector.count();
                 }
             }
+
+            @Override
+            public void visit(DocIdSetIterator iterator, byte[] packedValue) throws IOException {
+                long value = NumericUtils.sortableBytesToLong(packedValue, 0);
+                if (value > collector.activeRange[1]) {
+                    collector.finalizePreviousRange();
+
+                    if (collector.iterateRangeEnd(value)) {
+                        throw new CollectionTerminatedException();
+                    }
+                }
+                if (collector.activeRange[0] <= value && value <= collector.activeRange[1]) {
+                    for (int doc = iterator.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = iterator.nextDoc()) {
+                        collector.count();
+                    }
+                }
+            }
```

Comment:
Can reuse most of the code between both the visit methods, except for the iterator part

Reply:
Updated